### ISSUE TITLE
Can ACL softmax support  any  wxhxch size?

### ISSUE THREAD

**Issue by joey2014** [ARM-software/ComputeLibrary#79]
From ACL softmax source code, it seems to just support 1D? Can ACL support other dim softmax, for example, w x h x ch(w ,h and ch can be any number)?

seems channels must be 1:
terminate called after throwing an instance of 'std::runtime_error'
  what():  in void arm_compute::NESoftmaxLayer::configure(arm_compute::ITensor*, arm_compute::ITensor*) src/runtime/NEON/functions/NESoftmaxLayer.cpp:41: Number of channels 2. Required number of channels 1

**Comment by AnthonyARM** [comment]
The assert means the channels must be stored in a planar way rather than interleaved that's all. You can process tensors of an arbitrary number of dimensions as long as the data is stored in plans. 
As far as we know no network uses interleaved data but if you think this would be useful please provide an example use case.

**Comment by GeorgeARM** [comment]
Hello @joey2014,
Concerning the softmax layer, yes we currently support only 1D softmax.
Do your models require softmax to be applied on other dimensions?

**Comment by joey2014** [comment]
Hi @GeorgeARM ,
Yes, in MTCNN model, it uses 2D softmax, it likes 49x91x2(wxhxch). Caffe can support such 2D softmax.

**Comment by GeorgeARM** [comment]
Ok @joey2014 if this is something needed we can put it in the pipeline to implement. I will mark this as issue as feature request.

**Comment by GeorgeARM** [comment]
This is [fixed](https://git.mlplatform.org/ml/ComputeLibrary.git/commit/?h=refs/heads/master&id=efbf6c8fd54159b26eda43eea7a12fce491ca13a), you can test on latest master [here](https://review.mlplatform.org/#/admin/projects/ml/ComputeLibrary). Reopen if needed

**Comment by ssj4gogeta** [comment]
Hello @GeorgeARM, my data is in NCHW format and I am trying to calculate channelwise softmax after applying the above patch. I am giving axis parameter as 2, as per the instructions in that patch, but my answer is not matching with the values I manually compute. I even tried all other axis values just to be sure. But in none of the cases, the answer is matching.

My input dimensions are 1[batch size] * 2[channels] * 256[height] * 256[weight].
The output I want is e^{input[i]) / [ e^(input[i]) + e^(input[i+256*256]) ] for i <256*256 & 
e^{input[i]) / [ e^(input[i])+e^(input[i-256*256]) ] for i>256*256.
Basically calculating channelwise softmax as in Caffe.

Am I missing something? Any leads would be appreciated.
Thanks.

**Comment by GeorgeARM** [comment]
Hello @ssj4gogeta,
Well from what I gather you want to perform reduction on a single given axis. This issue is about supporting the [caffe/caffe2 softmax layer](https://caffe2.ai/docs/operators-catalogue.html#softmax) which has different behavior from what you expect.
Here the axis is used to coerce the input tensor to a 2D tensor and perform softmax.
What you want relates to #401

**Comment by ssj4gogeta** [comment]
I will check that and get back if there are any issues.
Thanks.
